### ISSUE TITLE
I2C feedback?

### ISSUE THREAD

**Issue by chrisspen** [Blueglide/ComMotion#1]
Hi, I also have a ComMotion that I'm trying to get working. Were you ever able to get encoder feedback via I2C?

Like your example, I was only able to get basic motor control working. I found [this demo code](http://letsmakerobots.com/files/ComMotion_Redboard_Demo.zip), which appears to show how to enable requesting encoder data via I2C, but I can't get it to work. When I call `SerialConfig()`, then I lose all control over ComMotion and my serial connection hangs.

**Comment by Blueglide** [comment]
Hello Chris.  No, I never went much beyond what I described on GirHub which only provided basic functionality. Never tried to use encoder feedback.

Unfortunately many potential useful ComMotion comments left by other enthusiasts on the letsmakerobots site have been removed. But I do have a note here that states that a user named "gadgetry" had reported that the board properly uses encoder feedback for motor speed control.

I wonder if there is any way to track him down?

Sorry I cannot be of more help!

**Comment by Blueglide** [comment]
Actually Chris, it appears that references to  gadgetry abound on the www.letsmakerobots .com site. Enter his name as a search term at the site, and you will get several pages of hits.  I'll bet if you contact him, you will be able to get some useful information regarding the ComMotion controller.

Let me know if you find out anything interesting.

**Comment by chrisspen** [comment]
Thanks for the info. Are you referring to [this guy](http://letsmakerobots.com/blog/23791)? Unfortunately, that site doesn't provide any way to contact users. I can't even seem to login. I just created an account but it's "pending approval".

I've confirmed the board uses the encoders in it's PID loop to maintain speed even when it encounters resistance, but I'd like to track that externally so I can estimate distance travelled.

**Comment by Blueglide** [comment]
Yes, that is the right guy.  I cannot remember his postings word-for-word but I suspect he only got as far as you already are - verifying that encoder feedback IS used in the PID speed control algorithm but the firmware not yet being developed to the point of making that data available to the user over the I2C link.

I see now that Sparkfun apparently is wise to the situation and the ComMotion board is no longer available from them. 

I suppose what is required is that some user or group of users take on the challenge of obtaining and enhancing the board's firmware to the point that some of this interesting data available to at the board is also available to user-developed Arduino software over the I2C link.  Not likely, I know.....

Good luck with your project!

**Comment by chrisspen** [comment]
They've published the firmware, which I have. It looks like they've
implemented it, but I can't be sure there isn't a bug. I'm just want to
confirm I'm not doing something dumb.

**Comment by chrisspen** [comment]
Hi,

I'm not sure if you still care about this, but I finally had some time to
research the problem reading encoder counts, and I figured out the problem.

It looks like the firmware was originally written to push the encoder
counts and other status data to the master, but this requires setting an
address on the master, sending that address to the ComMotion, and using the
correct clock speed between them (which appears to default to a
non-standard 400kHz on the ComMotion). After days of tinkering, I couldn't
get this to work, but I saw the firmware had an unimplemented
Wire.onRequest() handler, so I researched polling it via
Wire.requestFrom(), and sure enough, that was almost trivial to implement.

I also noticed it only tracks encoder counts if the motors are moving,
since that's the only way to detect direction without a quadrature encoder,
so I implemented separate encoder counts that track absolute encoder pulses.

I've published my code in case you're interested:

https://github.com/chrisspen/commotion/blob/master/src/firmware/ComMotion_Shield_V2_3/I2C_Send.ino
https://github.com/chrisspen/commotion/blob/master/src/examples/encoder_reader/src/main.ino

**Comment by Blueglide** [comment]
Hi Chris. I am not working on it right now, but I am still interested!  Sounds like your research paid off. I'll take a look at the code.

I originally got involved in this by my grandson who was using the CommMotion controller in a robotics project. I'll make sure he is aware of your work.

Thanks for keeping me in the loop!

Ken

**Comment by chrisspen** [comment]
No problem. The code you posted helped me iron out a few bugs when I first
started playing around with the ComMotion, so I wanted to return the favor.

Since you mentioned robotics, I should also mention I'm using it in a
robotics project as well. I'm making the project 100% open source. Although
the software is in beta and only about 50% complete, the hardware is
complete.

https://github.com/chrisspen/homebot
